lib/ts/recipe/thirdparty/recipeImplementation.ts

```diff
@@ -238,18 +235,52 @@ export default function getRecipeImplementation(recipeId: string, appInfo: Norma
             };
         },
         generateStateToSendToOAuthProvider: function (): string {
-            return generateThirdPartyProviderState();
+            // Generate state using algorithm described in https://github.com/supertokens/supertokens-auth-react/issues/154#issue-796867579
+            return `${1e20}`.replace(/[018]/g, (c) =>
+                (parseInt(c) ^ (crypto.getRandomValues(new Uint8Array(1))[0] & (15 >> (parseInt(c) / 4)))).toString(16)
+            );
         },
-        verifyStateFromOAuthProvider: function (input: {
-            stateFromProvider: string | undefined;
-            stateFromStorage: string | undefined;
+        verifyAndGetStateOrThrowError: async function (input: {
+            stateFromAuthProvider: string | undefined;
+            stateObjectFromStorage: StateObject | undefined;
             userContext: any;
-        }): boolean {
-            if (input.stateFromStorage === undefined || input.stateFromProvider === undefined) {
-                return false;
+        }): Promise<StateObject> {
+            if (
+                input.stateObjectFromStorage === undefined ||
+                input.stateObjectFromStorage.stateForAuthProvider === undefined
+            ) {
+                throw new Error("No valid auth state present in session storage");
+            }
+
+            if (input.stateFromAuthProvider === undefined) {
+                throw new Error("No state recieved from auth provider");
+            }
+
+            if (input.stateObjectFromStorage.expiresAt < Date.now()) {
+                throw new Error("Auth state verification failed. The auth provider took too long to respond");
+            }
+
+            if (input.stateFromAuthProvider !== input.stateObjectFromStorage.stateForAuthProvider) {
+                throw new Error("Auth state verification failed. The auth provider responded with an invalid state");
+            }
+
+            return input.stateObjectFromStorage;
+        },
+
+        getAuthCodeFromURL: function (): string {
+            const code = getQueryParams("code");
+
+            if (code === undefined) {
+                throw new Error(
+                    "There is no 'code' present in query params and no 'authCode' was provided when calling signInUp"
+                );
```

Comment:
Do not throw an error, instead return an empty string, and the API will eventually throw an error